src/background/config_backend.js

```diff
@@ -0,0 +1,87 @@
+import Logger from "../logger.js";
+import DBUtils from "../utilities.js";
+
+const defaultConfig = {
+  displayWaveform: false
+};
+
+export default class ConfigBackend {
+  constructor() {
+    this.log = new Logger();
+    this.dbUtils = new DBUtils();
+
+    this.boundConnectionListenerCallback = ConfigBackend.connectionListenerCallback.bind(
+      this
+    );
+
+    this.boundPortListenerCallback = ConfigBackend.portListenerCallback.bind(
+      this
+    );
+  }
+
+  init() {
+    this.log.info("initializing ConfigBackend");
+
+    chrome.runtime.onConnect.addListener(this.boundConnectionListenerCallback);
+  }
+
+  static connectionListenerCallback(port) {
+    this.log.info("connection listener callback");
+    if (port.name !== "bandcamplabelview") {
+      this.log.error(
+        `Unexpected chrome.runtime.onConnect port name: ${port.name}`
+      );
+      return;
+    }
+
+    this.port = port;
+    this.port.onMessage.addListener(this.boundPortListenerCallback);
+  }
+
+  static async portListenerCallback(msg) {
+    this.log.info("port listener callback");
+
+    const db = await this.dbUtils.getDB("config");
+
+    this.setupDB(db);
+
+    if (msg.config) this.synchronizeConfig(db, msg.config); //todo: remove this, it's not used
+
+    if (msg.toggleWaveformDisplay) this.toggleWaveformDisplay(db);
+
+    if (msg.requestConfig) this.broadcastConfig(db);
+  }
+
+  async synchronizeConfig(db, config) {
+    let db_config = await db.get("config", "config");
+    const merged_config = ConfigBackend.mergeData(db_config, config);
```

Comment:
In functional reactive programming this is known as a "reducer". Is `mergeData` used anywhere else? Might make sense to `Object.assign()` right here instead of calling a backend service to perform a one-line method.

Reply:
@emcniece I struggled to test the call. It's super easy to sub out the static method but I don't know if I can swap out the instance of `Object.assign()` with a stub/fake/mock...

but I will check

Reply:
Do we need to test the behaviour of a native method?

Reply:
@emcniece no, I don't really mean testing the functionality of Object.assign() (although the method that wraps it is tested). But you do need to test that the method was called within the section of code that needs it.

Reply:
I'd say you want to check that the config was merged, not that `Object.assign()` was called 😆

Reply:
@emcniece this is the thing. From the perspective of the test, I don't care what `mergeData` does and I don't want to have it actually run.

What I want is to know that that method was called and I want to be able to return values from that fake object. This lets me change the functionality of the merge component without having to re-write tests for `synchronizeConfig` (making refactoring easier) since my tests no longer check if  `synchronizeConfig` merges two objects (since that is not its responsibility).

Reply:
It should be possible to mock anything, native `Object.assign()` included, but sometimes that stuff just doesn't work properly. This level of abstraction to `mergeData` seems unnecessary, if testing is the problem you already have full control over the data coming out of `mergeData` because you call `synchronizeConfig` with half of the config and stub `dbStub` with the other half. The config halves can be the same, or they can be different depending on whether you're testing the merge behaviour or not.

Doesn't hurt anything though, just extra brain load 🤷